File: tests/fixtures/python_basic/src/payments/service.py
from payments.gateway import PaymentGateway
from payments.request import PaymentRequest


class PaymentService:
    gateway = None

    def __init__(self, gateway: PaymentGateway):
        self.gateway = gateway

    def authorize(self, request):
        validate(request)
        return self.gateway.charge(request)


def validate(request):
    if request is None:
        raise ValueError("request")
    PaymentRequest(request.user_id, request.amount)

File: tests/fixtures/go_basic/service.go
package payments

import (
	"context"
	"errors"
)

type PaymentService struct {
	Gateway PaymentGateway
}

type Authorizer interface {
	Authorize(ctx context.Context, request PaymentRequest) (Receipt, error)
}

func NewPaymentService(gateway PaymentGateway) *PaymentService {
	return &PaymentService{Gateway: gateway}
}

func (s *PaymentService) Authorize(ctx context.Context, request PaymentRequest) (Receipt, error) {
	if request.Amount <= 0 {
		return Receipt{}, errors.New("request")
	}
	s.validate(request)
	return s.Gateway.Charge(ctx, request)
}

func (s *PaymentService) validate(request PaymentRequest) error {
	if request.UserID == "" {
		return errors.New("user")
	}
	return nil
}
